core/include/drivers/i2c.h

```diff
@@ -0,0 +1,346 @@
+ *	Any TEE_Result compliant code in case of error.
+ *
+ * Returns a struct i2c pointer pointing to a reset controller matching
+ * the devicetree description or NULL if invalid description in which case
+ * @res provides the error code.
+ */
+typedef struct i2c_dev *(*i2c_dt_get_func)(struct dt_driver_phandle_args *a,
+					   void *data, TEE_Result *res);
+
+/**
+ * i2c_dt_register_provider - Register a reset controller provider
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the reset controller
+ * @get_dt_i2c: Callback to match the reset controller with a struct i2c
+ * @data: Data which will be passed to the get_dt_i2c callback
+ * Returns TEE_Result value
+ */
+static inline
+TEE_Result i2c_register_provider(const void *fdt, int nodeoffset,
+				 i2c_dt_get_func get_dt_i2c, void *data)
+{
+	int subnode = -1;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = dt_driver_register_provider(fdt, nodeoffset,
+					  (get_of_device_func)get_dt_i2c,
+					  data, DT_DRIVER_I2C);
+	if (res)
+		return res;
+
+	fdt_for_each_subnode(subnode, fdt, nodeoffset)
+		dt_driver_maybe_add_probe_node(fdt, subnode);
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * i2c_smbus_read_byte() - Execute a read byte SMBus protocol operation
+ *
+ * @i2c_dev: I2C device used for SMBus operation
+ * @cmd_code: Command code to read
+ * @byte: Returned byte value read from device
+ *
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result i2c_smbus_read_byte(struct i2c_dev *i2c_dev,
+					     uint8_t cmd_code, uint8_t *byte)
+{
+	return i2c_smbus_raw(i2c_dev, I2C_SMBUS_READ, I2C_SMBUS_PROTO_BYTE,
+			     cmd_code, byte, 1);
+}
+
+/**
+ * i2c_smbus_write_byte() - Execute a write byte SMBus protocol operation
+ *
+ * @i2c_dev: I2C device used for SMBus operation
+ * @cmd_code: Command code for write operation
+ * @byte: Byte to be written to the device
+ *
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result i2c_smbus_write_byte(struct i2c_dev *i2c_dev,
+					      uint8_t cmd_code, uint8_t byte)
+{
+	return i2c_smbus_raw(i2c_dev, I2C_SMBUS_WRITE, I2C_SMBUS_PROTO_BYTE,
+			     cmd_code, &byte, 1);
+}
+
+/**
+ * i2c_smbus_read_block_raw() - Execute a non-standard SMBus raw block read.
+ * This does not insert the "count" of byte to be written unlike the SMBus block
+ * read operation.
+ *
+ * @i2c_dev: I2C device used for SMBus operation
+ * @cmd_code: Command code for read operation
+ * @buf: Buffer of data read from device
+ * @len: Length of data to be read from the device
+ *
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result i2c_smbus_read_block_raw(struct i2c_dev *i2c_dev,
+						  uint8_t cmd_code,
+						  uint8_t *buf,
+						  unsigned int len)
+{
+	return i2c_smbus_raw(i2c_dev, I2C_SMBUS_READ, I2C_SMBUS_PROTO_BLOCK_RAW,
+			     cmd_code, buf, len);
+}
+
+/**
+ * i2c_smbus_write_block_raw() - Execute a non-standard SMBus raw block write.
```

Comment:
Since it's a non standard SMBus operation, maybe don't label this function  `i2c_smbus_xxx()` but rather `i2c_bus_write_block()`?

Reply:
yes, looks better.